src/end_entity.rs

```diff
@@ -185,4 +194,86 @@ impl<'a> EndEntityCert<'a> {
     pub fn dns_names(&'a self) -> Result<impl Iterator<Item = GeneralDnsNameRef<'a>>, Error> {
         subject_name::list_cert_dns_names(self)
     }
+
+    /// This function tries to verify that the DER encoded `private_key_der` bytes correspond
+    /// to the public key described in the end-entity certificate's subject public key information.
+    /// If the provided key isn't usable for this EE certificate [`Error::CertPrivateKeyMismatch`]
+    /// will be returned.
+    ///
+    /// This function supports the following private key algorithms and encodings (matching the
+    /// supported formats used by Rustls):
+    /// Key algorithms:
+    ///   * RSA
+    ///   * ECDSA (P-256 or P-384)
+    ///   * Ed25519
+    /// Encodings:
+    ///   * PKCS8v1 (RSA, ECDSA, Ed25519)
+    ///   * PKCS8v2 (Ed25519 only)
+    ///   * PKCS1 (RSA only)
+    ///   * Sec1 (ECDSA only)
+    pub fn verify_private_key(&self, private_key_der: &[u8]) -> Result<(), Error> {
+        // Parse the SPKI of the EE cert and extract the DER encoded bytes of the public key.
+        let spki = parse_spki_value(self.inner.spki.value())?;
+        let pubkey = spki.key_value.as_slice_less_safe();
+
+        fn matches_or_error(pubkey: &[u8], pubkey_tbv: &[u8]) -> Result<(), Error> {
+            match pubkey == pubkey_tbv {
+                true => Ok(()),
+                false => Err(Error::CertPrivateKeyMismatch),
+            }
+        }
```

Comment:
I would probably extract this as a free function. I also think this could take a `K: KeyPair` instead of taking the `pubkey` directly. You could even go all in on this and abstract towards an `impl Fn(&[u8]) -> Option<impl KeyPair>` to reduce duplication more.

(Style nit: despite existing code, I think `pub_key` would be more idiomatic spelling in Rust than `pubkey`. It's also not obvious to me what `_tbv` means. Maybe something more like `cert_pub_key` and `pair_pub_key`?)

Reply:
Reworked :+1: